src/main/java/org/openstreetmap/atlas/geography/atlas/delta/Diff.java

```diff
@@ -100,23 +100,55 @@ public static String toGeoJson(final Iterable<Diff> diffs)
      */
     public static String toGeoJson(final Iterable<Diff> diffs, final Predicate<Diff> filter)
     {
-        return new GeoJsonBuilder().create(Iterables.stream(diffs)
-                .filter(diff -> diff.getItemType() != ItemType.RELATION).filter(filter).map(diff ->
-                {
-                    final AtlasItem item;
-                    if (diff.getDiffType() == DiffType.REMOVED)
-                    {
-                        item = (AtlasItem) diff.getBaseEntity();
-                    }
-                    else
-                    {
-                        item = (AtlasItem) diff.getAlterEntity();
-                    }
-                    final Map<String, String> tags = item.getTags();
-                    tags.put("DIFF_TYPE", diff.getDiffType().name());
-                    tags.put("DIFF_REASON", diff.getDiffReason().name());
-                    return new LocationIterableProperties(item.getRawGeometry(), tags);
-                }).collect()).jsonObject().toString();
+        return new GeoJsonBuilder().create(
+                Iterables.stream(diffs).filter(diff -> diff.getItemType() != ItemType.RELATION)
+                        .filter(filter).flatMap(Diff::processDiff).collect())
+                .jsonObject().toString();
+    }
+
+    private static List<LocationIterableProperties> processDiff(final Diff diff)
+    {
+        final List<LocationIterableProperties> items = new ArrayList<>();
+
+        final AtlasEntity beforeEntity = diff.getBeforeEntity();
+        // null if it is ADDED
+        if (beforeEntity != null)
+        {
+            final Map<String, String> beforeTags = beforeEntity.getTags();
+            beforeTags.put("diff", "BEFORE");
+            beforeTags.put("diff:type", diff.getDiffType().name());
+            beforeTags.put("diff:reason", diff.getDiffReason().name());
+            if (diff.getItemType() == ItemType.RELATION)
+            {
+                items.addAll(processRelationForGeoJson((Relation) beforeEntity, beforeTags));
+            }
+            else
+            {
+                items.add(new LocationIterableProperties(
+                        ((AtlasItem) beforeEntity).getRawGeometry(), beforeTags));
+            }
+        }
+
+        final AtlasEntity afterEntity = diff.getAfterEntity();
+        // null iff it is REMOVED
+        if (afterEntity != null)
+        {
+            final Map<String, String> afterTags = afterEntity.getTags();
+            afterTags.put("diff", "AFTER");
```

Comment:
It seems like this is the only difference between these code blocks. Maybe could move the if(null) check here and avoid duplicating a lot of this code?

Reply:
Great point, I will change this.